### PR TITLE
Add error message + code to failing bulk operations on the HTTP interface

### LINKED ISSUE #16649: Show error messages in result array for failed bulk inserts via HTTP endpoint
### Problem Statement

We are inserting many rows via python, i.e. the HTTP interface of CrateDB.

[Since version 4.2](https://github.com/crate/crate/blob/54105287c92be84a29d223e18a7e29ef676e7ae3/docs/appendices/release-notes/4.2.0.rst?plain=1#L53-L56), it reports the rowcount in the results array, either 1 for success, or -2 in case of failure and does not show an error.

In order to avoid single inserts to be executed to get the particular error message for the failed inserts as outlined in https://github.com/crate/cratedb-toolkit/pull/262, it would be great to see the particular error code and exception message directly in the results array (not a full stack trace as the [error_trace=true](https://cratedb.com/docs/crate/reference/en/latest/interfaces/http.html#error-handling) parameter would enable).

**Current output:**
```
{
    "cols": [],
    "duration": 31.547579,
    "results": [
        {
            "rowcount": 1
        },
        {
            "rowcount": -2
        },
        {
            "rowcount": -2
        }
    ]
}
```

**Ideal output:**
```
{
    "cols": [],
    "duration": 31.547579,
    "results": [
        {
            "rowcount": 1
        },
        {
            "rowcount": -2,
            "error": 4091,
            "message": "DuplicateKeyException[A document with the same primary key exists already]"
        },
        {
            "rowcount": -2,
            "error": 4000,
            "message": "SQLParseException[Cannot cast object element `a` with value `asdf` to type `bigint`]"
        }
    ]
}
```

Data to reproduce the ideally expected result above:

```
CREATE TABLE insert_test (id TEXT PRIMARY KEY, val OBJECT(DYNAMIC));

-- seed table
INSERT INTO insert_test(id, val) VALUES ('1', '{ "a": 1, "b": { "c": 1}, "d": [{"k": "k1", "v": 1}]}');

-- insert works, no error to report
INSERT INTO insert_test(id, val) VALUES ('2', '{ "a": 2, "b": { "c": 2}, "d": [{"k": "k1", "v": 2}]}');

-- DuplicateKeyException[A document with the same primary key exists already]
INSERT INTO insert_test(id, val) VALUES ('2', '{ "a": 2, "b": { "c": 2}, "d": [{"k": "k1", "v": 2}]}');

-- SQLParseException[Cannot cast object element `a` with value `asdf` to type `bigint`]
INSERT INTO insert_test(id, val) VALUES ('3', '{ "a": "asdf", "b": { "c": 3}, "d": [{"k": "k1", "v": 3}]}');
```

HTTP bulk insert of the above statements:
```
curl -sS -H 'Content-Type: application/json' \
-X POST '127.0.0.1:4200_sql' -d@- <<- EOF
{
  "stmt": "INSERT INTO insert_test(id, val) VALUES (?, ?)",
  "bulk_args": [
    ["2", "{ \"a\": 2, \"b\": { \"c\": 2}, \"d\": [{\"k\": \"k1\", \"v\": 2}]}"],
    ["2", "{ \"a\": 2, \"b\": { \"c\": 2}, \"d\": [{\"k\": \"k1\", \"v\": 2}]}"],
    ["3", "{ \"a\": \"asdf\", \"b\": { \"c\": 3}, \"d\": [{\"k\": \"k1\", \"v\": 3}]}"]
  ]
}
EOF
```

### Possible Solutions

If we don't want to change the default behaviour, maybe a switch to enable error reporting can be introduced.

### Considered Alternatives

An alternative would be to execute single inserts for each failed record as done in https://github.com/crate/cratedb-toolkit/pull/262, but when processing a lot of records that are failing, this leads to many single inserts significantly reducing the performance when required to be executed for 1000's of records.

### PR DESCRIPTION
**Pull request by seut**
For each REST bulk response entry, an `error` object containing a `message` and `code` is added in case of an exception.
Each involving shard will return at most `10` error details. If more errors occur, the corresponding bulk response entry will still only contain the `-2` row count without further details. This is to avoid too much memory consumption by errors which could lead to unwanted CircuitBreaker exceptions only caused by trying to transfer the full error exception. This isn't wanted and also one might mostly be interested in the most X errors.

Closes #16649.

### CODE REVIEW

File: server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.dml;
+
+import org.jetbrains.annotations.Nullable;
+
+import io.crate.data.Row1;
+
+public final class RowCountAndFailure {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java:28]: I know PR is still in draft but drive-by comment anyway:

Maybe worth to go with a struct of arrays instead of array of structs here and have something like `Rowcounts` / `BulkResponse` that has `long[] counts` and `Throwable[] failures` for a smaller memory footprint.

And I think we should look into limiting the amount of `Throwable` instances we keep - e.g. maybe only keep up to 50 or 100 errors, and discard afterwards.
E.g. with a bulk size of 10_000 and each item failiing, and each having a stack of somewhere of 50-100 we could end up using lots of memory and running into circuit breaker/OOM because of the failure tracking.

Looks like `Throwable` is used for the permission checks - but maybe we could even do these eagerly and only keep the message as string?

Comment (seut) [reply server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java:28]: Thanks for the feedback, valid points, will look into them.

> Looks like Throwable is used for the permission checks - but maybe we could even do these eagerly and only keep the message as string?

Not only permission checks, but also to convert them into user friendly exceptions.
In order to move the permission check on the shard, we'd need to stream the `CoordinatorSessionSettings` (so they wouldn't be *Coordinator*  once anymore). But we probably need to do this anyhow, to fix information leaks already existing when using `RETURN SUMMARY`, see https://github.com/crate/crate/pull/16799/files#diff-ccd3239453fb803999cc796e8f97365611ff44ead633df9db0acdeda147bb3b4R433.

Comment (seut) [reply server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java:28]: Another issue with converting the `Throwable` directly into a message, is the error `code`. This code is very interface related, we currently generate it out of the `Throwable` only on the REST interface. The PG interface may create different error codes. I see these options when we want to avoid streaming the exception:
- a) Create the error code based on the error message   <- sounds very error-prone
- b) Create the error codes at the shards and transport them <- moves interface(API) related code deep into the execution
- c) Create internal codes we can use to create API specific codes
- d) ?

I'd favor c).

Comment (seut) [reply server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java:28]: For c) we may could reuse the already existing exception id's of `ElasticsearchExceptionHandle`.

Comment (seut) [reply server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java:28]: I've pushed a rough draft implementation of c) by https://github.com/crate/crate/pull/16799/commits/0ca94ee0d9b6e4f003489c864003a5a07a7185ef, please also check this commit comments.

Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/dml/RowCountAndFailure.java:28]: I'm not sure either about which way to go but I'd almost tend to keep the exceptions but ensure we never keep more than ~50 or so in memory.
And also change the existing ShardResponse.Failure to work with exceptions instead of messages.

The main reason is that converting exceptions just before they're sent to the client to account for privileges seems like an easier model to reason about than if the logic gets done in different phases.

And keeping the original exceptions around for longer might help with debugging during development.

File: server/src/main/java/io/crate/execution/dml/delete/TransportShardDeleteAction.java [hunk 2]
```diff
@@ -103,21 +106,23 @@ protected WritePrimaryResult<ShardDeleteRequest, ShardResponse> processRequestIt
...
-                            new ShardResponse.Failure(
-                                item.id(),
-                                "Document not found while deleting",
-                                false));
+                        errorCount++;
+                        Throwable ex = null;
+                        if (errorCount <= BULK_RESPONSE_MAX_ERRORS_PER_SHARD) {
+                            ex = new DocumentMissingException(indexShard.shardId(), Constants.DEFAULT_MAPPING_TYPE, item.id());
+                        }
+                        shardResponse.add(location, new ShardResponse.Failure(item.id(), ex, false));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/dml/delete/TransportShardDeleteAction.java:114]: If I'm reading this right maybe the `errorCount` & `<=   BULK_RESPONSE_MAX_ERRORS_PER_SHARD`  handling could be moved into `shardResponse` to avoid some duplication between shardDeleteAction and shardUpsertAction and encapsulate it.

Not sure if worth it, but then we could even follow up in another PR to do some sort of de-duduplication - instead of adding new failure instances we keep a list of failed item id's with the same error or something like that.

File: server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java [hunk 3]
```diff
@@ -41,7 +42,7 @@ public final class ShardedRequests<TReq extends ShardRequest<TReq, TItem>, TItem
     final Map<String, List<ReadFailureAndLineNumber>> itemsWithFailureBySourceUri = new HashMap<>();
     final Map<String, String> sourceUrisWithFailure = new HashMap<>();
     final List<RowSourceInfo> rowSourceInfos = new ArrayList<>();
-    final Map<ShardLocation, TReq> itemsByShard = new HashMap<>();
+    final Map<ShardLocation, TReq> itemsByShard = new LinkedHashMap<>();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java:45]: What's the reason for this change? Increases memory use

File: server/src/main/java/io/crate/rest/action/HttpError.java [hunk 4]
```diff
@@ -151,6 +152,8 @@ public static HttpError fromThrowable(Throwable throwable) {
             httpErrorStatus = HttpErrorStatus.SNAPSHOT_WITH_SAME_NAME_EXISTS_ALREADY;
         } else if (throwable instanceof UnavailableShardsException) {
             httpErrorStatus = HttpErrorStatus.ONE_OR_MORE_SHARDS_NOT_AVAILABLE;
+        } else if (throwable instanceof DocumentMissingException) {
+            httpErrorStatus = HttpErrorStatus.DOCUMENT_NOT_FOUND;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/rest/action/HttpError.java:156]: Newer pattern is to implement `httpErrorStatus` on the exception if it's a `ElasticsearchException`.
We should probably add a comment about that within this method. The same is also the case for the PG error code.

**Review by mfussenegger — APPROVED** [review]
Nice. Left a couple of minor commets

File: server/src/main/java/io/crate/rest/action/HttpError.java (hunk 4 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/rest/action/HttpError.java:156]: Looks like as a follow up PR we could also remove `RestStatus status` it's not used in a meaningful way and overlaps with the more detailed `HttpErrorStatus`

Comment (seut) [reply server/src/main/java/io/crate/rest/action/HttpError.java:156]: Ah thanks, will change.

File: server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java (hunk 3 above)
Comment (seut) [reply server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java:45]: Not sure anymore why I did this, seems not needed, I'll revert. Thanks for catching!

File: server/src/main/java/io/crate/execution/dml/delete/TransportShardDeleteAction.java (hunk 2 above)
Comment (seut) [reply server/src/main/java/io/crate/execution/dml/delete/TransportShardDeleteAction.java]: Good point, done via https://github.com/crate/crate/pull/16799/commits/1a186eb52a8c90b27155ce0ff443e490958d41d7

File: server/src/main/java/io/crate/rest/action/HttpError.java (hunk 4 above)
Comment (seut) [reply server/src/main/java/io/crate/rest/action/HttpError.java]: Followup work #17206.
